src/query/sql/src/planner/binder/location.rs

```diff
@@ -44,6 +48,91 @@ fn secure_omission(endpoint: String) -> String {
     }
 }
 
+static SCHEME_PARAMS: OnceCell<HashMap<Scheme, HashSet<String>>> = OnceCell::new();
+
+fn check_uri(protocol: &Scheme, conns: &BTreeMap<String, String>) -> Result<()> {
```

Comment:
Compared to a general `check_uri`, how about implementing functions like `parse_s3_uri_location`? In which we will do `parse_s3_uri_location(l: UriLocation) -> Result<StorageParams>`.

Implementing `parse_xxx_uri_location` for every scheme could produce more code but will

- make future maintaining easier (all code related to this scheme belongs here).
- Remove the need of `HashMap<Scheme, HashSet<String>>`, we can have a static hashset for every scheme.
- Make unit tests for easier.

Reply:
I initially considered this approach, but

1. it would generate a lot of code
2. there would be a lot of global HashSet

Considering that the number of parameters for each cloud platform is not very large, I finally took a unified approach.

Reply:
We do these checks in a dynamic way.

For example, build a `ConnectionChecker` which provides the following API:

- `get(key) -> Value`: store the key in a visited `HashSet` and than fecth value from `conn` HashMap.
- `check() -> Result<()>`: after all keys have been checked, we can compared the visited hashset with user input's `conn`, if user inputs have the key that not in `visisted`, we can return an error with UnknownKeyInputted.

In this way, we don't need to maintain a static hashset for every services.

Reply:
If we split up the functions by scheme, it would result in a lot of duplicate code, and the logic would be almost identical except construction of different StorageParams by scheme.

Reply:
My idea is 

![image](https://user-images.githubusercontent.com/5351546/210165995-ca88e2af-a0b6-4eda-9281-23ca3d28e9d0.png)